### PR TITLE
AOF channel

### LINKED ISSUE #130: AOF should be reading off a channel
Instead of allowing direct communication we need to have a channel open for AOF feeding

### PR DESCRIPTION
**Pull request by bahlo**
This pr fixes #130.

**Edit:** The following changed over the discussion of this pr, please see the final commit 591b6cd for more details.

## Added
- `writeChan` for write requests
- `writeReq` struct to contain entry to write and error chan
- `writeQueue` func, which adds the entry to the queue and returns an error chan
- Asynchronous go func which receives items from the writeChannel and writes 
  them in order to the file 
- `Stop` func which stops writing and waits on items in the channel to finish write
- Tests for `Stop` func

## Removed
- `lock` mutex

### CODE REVIEW

**Comment by bahlo** [comment]
Do you want the Append*Op methods to be asynchronous? I could easily just return the channel instead of waiting on the value and returning that.

Plus, do you want a channel for read values as well? And if yes, should there be only one receiver or rather pub/sub?

**Comment by seiflotfy** [comment]
you got it right afaict. Need to review properly. But the main idea is the server gets a write request and tells AOF to store it, once AOF tells it its done, it forwards it to the manager.

File: src/storage/aof.go [hunk 1]
```diff
@@ -62,13 +127,12 @@ func (aof *AOF) AppendDomOp(op uint8, dom proto.Message) error {
 		return fmt.Errorf("No such op %d", op)
 	}
 	e.op = op
-	return aof.write(e)
+
+	return <-aof.writeQueue(e)
 }
 
 // AppendSketchOp ...
 func (aof *AOF) AppendSketchOp(op uint8, sketch proto.Message) error {
```
Comment (seiflotfy) [inline src/storage/aof.go]: Would it make sense for the server to communicate with the channel directly instead over method?
@njpatel @bahlo @martinpinto what do you think?

Comment (bahlo) [inline src/storage/aof.go]: I thought about that, but the error get's checked on each occurrence of Append*Op in the server code, so I thought it'd be easier to just return the received value. Your call.

**Comment by bahlo** [comment]
#### Small update on why I chose this implementation:

My first thought was creating two channels, one for the write queue and one for the errors.
The problem with this is, that you can't know which write caused an error since there is no way to synchronise those channels without compromising speed.

It'd also be possible to just return the `<-chan error` from the Append*Op methods. This may be the better approach, it would create asynchronous methods since they return almost immediately and the `error`/`nil` value get's put on the channel after receive. You'd have to make it a buffered channel to make receiving optional (without blocking everything).

**Comment by seiflotfy** [comment]
So ideally we would have an AOF method called Run(incoming chan, outgoing chan) where in server its called via
```go 
go aof.Run(incoming, outgoing) // It might be better to set the channel during construction of the AOF
```

The incoming takes it "Entry" and the outgoing also delivers ONLY the successful "Entry"
Makes sense?

**Comment by bahlo** [comment]
I pushed b70eab3, which removes the `<-chan error` return value and ignores errors altogether (although we could create a channel per AOF which receives errors). 
A write now returns almost immediately and successful written entries are put on the writtenChan.

I prefer not giving any control to calling methods they don't need to have (like creating channels), but I'm wiling to add those parameters to `Run()` if you want that. :+1: 

#### Example
```go
// Create AOF
aof := NewAOF(...)

// Run writer method
go aof.Run()

// Log Op of written items
go func() {
  wc := aof.WrittenChan()
  for {
    if e, more := <-wc; more {
      // We have a new successful written Entry, print Op
      fmt.Println(e.Op())
    } else {
      break
    }
  }
}()

sketch := createSketch("skz1", pb.SketchType_CARD)
aof.AppendDomOp(aof.DeleteDom, sketch)

aof.Stop() // Stop the writer go routine

// Prints: 
// 1
```

File: src/server/domain.go [hunk 2]
```diff
@@ -54,9 +51,7 @@ func (s *serverStruct) deleteDomain(ctx context.Context, in *pb.Domain) (*pb.Emp
 }
 
 func (s *serverStruct) DeleteDomain(ctx context.Context, in *pb.Domain) (*pb.Empty, error) {
-	if err := s.storage.AppendDomOp(storage.DeleteDom, in); err != nil {
-		fmt.Println(err)
-	}
+	s.storage.AppendDomOp(storage.DeleteDom, in)
 	return s.deleteDomain(ctx, in)
```
Comment (seiflotfy) [inline src/server/domain.go]: This should only happen when being read off the channel

File: src/server/domain.go [hunk 3]
```diff
@@ -31,9 +30,7 @@ func (s *serverStruct) createDomain(ctx context.Context, in *pb.Domain) (*pb.Dom
 }
 
 func (s *serverStruct) CreateDomain(ctx context.Context, in *pb.Domain) (*pb.Domain, error) {
-	if err := s.storage.AppendDomOp(storage.CreateDom, in); err != nil {
-		fmt.Println(err)
-	}
+	s.storage.AppendDomOp(storage.CreateDom, in)
```
Comment (seiflotfy) [inline src/server/domain.go]: Remove line i guess

File: src/server/sketch.go [hunk 4]
```diff
@@ -19,9 +18,7 @@ func (s *serverStruct) createSketch(ctx context.Context, in *pb.Sketch) (*pb.Ske
 }
 
 func (s *serverStruct) CreateSketch(ctx context.Context, in *pb.Sketch) (*pb.Sketch, error) {
-	if err := s.storage.AppendSketchOp(storage.CreateSketch, in); err != nil {
-		fmt.Println(err)
-	}
+	s.storage.AppendSketchOp(storage.CreateSketch, in)
```
Comment (seiflotfy) [inline src/server/sketch.go]: Remove :)

File: src/server/sketch.go [hunk 5]
```diff
@@ -45,9 +42,7 @@ func (s *serverStruct) add(ctx context.Context, in *pb.AddRequest) (*pb.AddReply
 }
 
 func (s *serverStruct) Add(ctx context.Context, in *pb.AddRequest) (*pb.AddReply, error) {
-	if err := s.storage.AppendAddOp(in); err != nil {
-		fmt.Println(err)
-	}
+	s.storage.AppendAddOp(in)
```
Comment (seiflotfy) [inline src/server/sketch.go]: And again :D

File: src/server/sketch.go [hunk 6]
```diff
@@ -114,9 +109,7 @@ func (s *serverStruct) deleteSketch(ctx context.Context, in *pb.Sketch) (*pb.Emp
 }
 
 func (s *serverStruct) DeleteSketch(ctx context.Context, in *pb.Sketch) (*pb.Empty, error) {
-	if err := s.storage.AppendSketchOp(storage.DeleteSketch, in); err != nil {
-		fmt.Println(err)
-	}
+	s.storage.AppendSketchOp(storage.DeleteSketch, in)
```
Comment (seiflotfy) [inline src/server/sketch.go]: Last one :)

File: .gitignore [hunk 7]
```diff
@@ -28,3 +28,4 @@ _testmain.go
 bin
 pkg
 vendor/src
+test.log
```
Comment (martinpinto) [inline .gitignore]: This is no longer necessary as it was fixed by #138

**Comment by seiflotfy** [comment]
#139 fixes it

**Comment by bahlo** [comment]
@seiflotfy The implementation in #139 is the same principle like my first PR commit, but great this is now working :+1:

**Comment by seiflotfy** [comment]
Yeah I picked out some stuff and attributed it back to you :D

**Pull request closed without merging**